uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js

```diff
@@ -379,9 +397,37 @@ define([
      */
     function initListeners () {
       listeners.push(
-        $rootScope.$on('LeaveRequestPopup::updateBalance', function () {
-          vm.updateBalance();
-          setDaySelectionMode();
+        $rootScope.$on('LeaveRequestPopup::updateAbsenceType', function () {
+          var absenceTypeUnitChanged;
+          var previousAbsenceTypeUnit = vm.selectedAbsenceType.calculation_unit_name;
+
+          vm.selectedAbsenceType = getSelectedAbsenceType();
+          absenceTypeUnitChanged = previousAbsenceTypeUnit !== vm.selectedAbsenceType.calculation_unit_name;
+
+          setOpeningBalance();
+
+          return $q.resolve()
+            .then(absenceTypeUnitChanged && setDaysSelectionMode)
+            .then(function () {
+              if (absenceTypeUnitChanged && isCalculationUnit('hours')) {
+                return loadTimeRangesFromWorkPattern('from');
+              }
+            })
+            .then(setRequestDates)
+            .then(performBalanceChangeCalculation);
+        }),
+        $rootScope.$on('LeaveRequestPopup::updateBalance', function (event, absenceTypesWithBalances) {
```

Comment:
sorry but this still doesn't make sense. The absence types had already been updated with the new entitlements in the request controller
```js
// request.controller.js

$rootScope.$on('LeaveRequestPopup::absencePeriodChanged', function () {
  loadEntitlements()
    .then(setAbsenceTypesFromEntitlements) // <-- here
    .then(function () {
      $rootScope.$emit('LeaveRequestPopup::updateBalance', vm.absenceTypes);
    });
}),
```
and this component have a binding with those
```js
components.component('leaveRequestPopupDetailsTab', {
  bindings: {
    absenceTypes: '<',
```
so it should get them automatically without having them awkwardly passed as part of an event that has nothing to do with them

Reply:
But it doesn't. That's what happens if I simply remove the setter.

```js
function updateBalances (absenceTypesWithBalances) {
      //vm.absenceTypes = absenceTypesWithBalances; // <<<<<<<<<<<<

      setSelectedAbsenceType();

      if (moment(vm.uiOptions.toDate).isAfter(vm.period.end_date)) {
        vm.uiOptions.toDate = undefined;
      .... etc
```

![image](https://user-images.githubusercontent.com/3973243/34003491-4951851e-e0ed-11e7-9649-5b69722bf236.png)

It is a tick problem. It can be solved with $timeout:

```js
$rootScope.$on('LeaveRequestPopup::absencePeriodChanged', function () {
  loadEntitlements()
    .then(setAbsenceTypesFromEntitlements) // <-- here
    .then(function () {
      $.timeout(function () { // <<<<< here
        $rootScope.$emit('LeaveRequestPopup::updateBalance', vm.absenceTypes);
      });
    });
}),
```

I genuinely don't know how to solve this either way.